google/cloud/bigtable/mutations.py

```diff
@@ -13,42 +13,121 @@
 # limitations under the License.
 #
 from __future__ import annotations
-
+from typing import Any
 from dataclasses import dataclass
-from google.cloud.bigtable.row import family_id, qualifier, row_key
+from abc import ABC, abstractmethod
+
+
+class Mutation(ABC):
+    """Model class for mutations"""
 
+    @abstractmethod
+    def _to_dict(self) -> dict[str, Any]:
+        raise NotImplementedError
 
-class Mutation:
-    pass
+    def is_idempotent(self) -> bool:
+        """
+        Check if the mutation is idempotent
+        If false, the mutation will not be retried
+        """
+        return True
+
+    def __str__(self) -> str:
+        return str(self._to_dict())
 
 
 @dataclass
 class SetCell(Mutation):
-    family: family_id
-    column_qualifier: qualifier
-    new_value: bytes | str | int
-    timestamp_ms: int | None = None
+    family: str
+    qualifier: bytes
+    new_value: bytes
+    timestamp_micros: int | None = None
+
+    def _to_dict(self) -> dict[str, Any]:
+        """Convert the mutation to a dictionary representation"""
+        # if timestamp not given, use -1 for server-side timestamp
+        timestamp = self.timestamp_micros if self.timestamp_micros is not None else -1
```

Comment:
I don't think we should implicitly set it to use server timestamps if this field is not set. If it's unset we can use the current timestamp. (java implementation: https://github.com/googleapis/java-bigtable/blob/main/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/Mutation.java#LL155C1-L156C1). Implicitly using server side timestamp could lead to duplicate mutations.

Reply:
Sounds good, I made the change to match Java